lib/maths/CBoostedTreeImpl.cc

```diff
@@ -1445,5 +1446,54 @@ const double CBoostedTreeImpl::INF{std::numeric_limits<double>::max()};
 const CBoostedTreeHyperparameters& CBoostedTreeImpl::bestHyperparameters() const {
     return m_BestHyperparameters;
 }
+
+void CBoostedTreeImpl::computeShapValues(int topShapValues,
+                                         core::CDataFrame& frame,
+                                         const TProgressCallback&) {
+    if (m_BestForestTestLoss == INF) {
+        HANDLE_FATAL(<< "Internal error: no model available for prediction. "
+                     << "Please report this problem.");
+        return;
+    }
+    bool successful;
+    auto treeFeatureImportance =
+        std::make_unique<maths::CTreeShapFeatureImportance>(m_BestForest);
+    TDoubleVecVec shapValues;
+    TDoubleVec shapTotal;
+    size_t numberInputFields = predictionColumn(frame.numberColumns()) - 1;
+    topShapValues = (topShapValues < numberInputFields) ? topShapValues : numberInputFields;
+    std::tie(shapValues, shapTotal) =
+        treeFeatureImportance->shap(frame, *m_Encoder, numberInputFields);
+
+    // get indices of the top elements
+    std::vector<std::size_t> indices(shapTotal.size());
+    std::iota(indices.begin(), indices.end(), 0);
+    std::nth_element(indices.begin(), indices.end(), indices.begin() + topShapValues - 1,
+                     [&shapTotal](std::size_t a, std::size_t b) {
+                         return shapTotal[a] > shapTotal[b];
+                     });
+
+    // write top columns to the result
+    std::size_t offset{frame.numberColumns()};
+    frame.resizeColumns(m_NumberThreads, frame.numberColumns() + topShapValues);
+    TStrVec columnNames(frame.columnNames());
+    for (int i = 0; i < topShapValues; ++i) {
+        columnNames[offset + i] = SHAP_PREFIX + frame.columnNames()[indices[i]];
+    }
+    frame.columnNames(columnNames);
+
+    std::tie(std::ignore, successful) = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(), [&](TRowItr beginRows, TRowItr endRows) {
+            for (auto row = beginRows; row != endRows; ++row) {
+                for (int i = 0; i < topShapValues; ++i) {
```

Comment:
nit: can we make loops over `std::size_t` in this function.

Reply:
Following your suggestion, I removed the need for the loops here altogether.